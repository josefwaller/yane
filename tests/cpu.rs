mod common;
use log::*;
use std::{
    cmp::min,
    fs::{read, File},
    io::{BufRead, BufReader, Read},
    string,
};

use assert_hex::assert_eq_hex;
use yane::{Cartridge, Controller, Nes};

// Runs the NES test CPU file and checks the state of the NES after each execution
#[test]
fn test_nestest_log() {
    let rom = read("./tests/test_roms/cpu_nestest.nes").unwrap();
    let mut nes = Nes::from_cartridge(Cartridge::new(&rom, None));
    nes.cpu.p_c = 0xC000;
    let f = File::open("./tests/nestest.log").unwrap();
    let buf = BufReader::new(f);
    let mut cycles: i64 = 7;
    for line in buf.lines() {
        let l = line.unwrap();
        macro_rules! assert_eq_substr {
            ($left: expr, $start: expr, $end: expr, $name: expr) => {
                assert_eq_hex!(
                    $left as u16,
                    get_hex_u16(&l, $start, $end),
                    "Expected {} to be {:X}, but was {:X}",
                    $name,
                    get_hex_u16(&l, $start, $end),
                    $left,
                );
            };
        }
        println!("{}", l);
        assert_eq_substr!(nes.cpu.p_c, 0, 4, "PC");
        assert_eq_substr!(nes.cpu.a, 50, 2, "A");
        assert_eq_substr!(nes.cpu.x, 55, 2, "X");
        assert_eq_substr!(nes.cpu.y, 60, 2, "Y");
        assert_eq_substr!(nes.cpu.s_r.to_byte(), 65, 2, "SR");
        assert_eq_substr!(nes.cpu.s_p, 71, 2, "SP");
        let c = l[90..].parse::<i64>().unwrap();
        assert_eq!(cycles, c, "Cycles should be {}, is {}", c, cycles);
        println!(
            "Executing {:#X} {:#X} {:#X} at PC = {:#X} (A = {:X}, X = {:X}, Y = {:X}, SR = {:X} = {:b})",
            nes.read_byte(nes.cpu.p_c as usize),
            nes.read_byte(nes.cpu.p_c as usize + 1),
            nes.read_byte(nes.cpu.p_c as usize + 2),
            nes.cpu.p_c,
            nes.cpu.a,
            nes.cpu.x,
            nes.cpu.y,
            nes.cpu.s_r.to_byte(),
            nes.cpu.s_r.to_byte()
        );
        match nes.step() {
            Ok(c) => cycles += c as i64,
            Err(s) => panic!("{}", s),
        }
    }
}

fn get_hex(s: &String, start: usize, len: usize) -> u8 {
    get_hex_u16(s, start, len) as u8
}
fn get_hex_u16(s: &String, start: usize, len: usize) -> u16 {
    u16::from_str_radix(&s[start..min(start + len, s.len())], 16).unwrap()
}

// Check the background to verify the NES test results
#[test]
fn test_nestest_file() {
    let mut nes = nes_with_rom!("./test_roms/cpu_nestest.nes");
    advance_nes_frames!(nes, 100);
    press_button!(nes, 0, start);
    advance_nes_frames!(nes, 100);
    release_button!(nes, 0, start);
    assert_background_snapshot!("nestest_official", nes);
    press_button!(nes, 0, select);
    advance_nes_frames!(nes, 100);
    press_button!(nes, 0, start);
    advance_nes_frames!(nes, 100);
    assert_background_snapshot!("nestest_unofficial", nes);
}
#[test]
fn test_official() {
    rom_test!("./test_roms/cpu_official.nes", 8_000);
}

#[test]
fn test_branch() {
    rom_test!("./test_roms/cpu_branch.nes");
}
