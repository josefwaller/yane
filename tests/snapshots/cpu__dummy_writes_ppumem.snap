---
source: tests/cpu.rs
assertion_line: 98
expression: screen.as_slice()
snapshot_kind: text
---
[
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0 34 25 33 34 1A  0 43 50 55 3F 44 55 4D 4D 59 3F 57 52 49 54 45 53 3F 50 50 55 4D 45 4D  0  0",
    " 0 94 A8 A9 B3 60 B0 B2 AF A7 B2 A1 AD 60 B6 A5 B2 A9 A6 A9 A5 B3 60 B4 A8 A1 B4 60 B4 A8 A5  0",
    " 0 83 90 95 60 A4 AF A5 B3 60 72 B8 60 B7 B2 A9 B4 A5 B3 60 B0 B2 AF B0 A5 B2 AC B9 6E  0  0  0",
    " 0 81 AE B9 60 B2 A5 A1 A4 6D AD AF A4 A9 A6 B9 6D B7 B2 A9 B4 A5 60 AF B0 A3 AF A4 A5  0  0  0",
    " 0 B3 A8 AF B5 AC A4 60 A6 A9 B2 B3 B4 60 B7 B2 A9 B4 A5 60 B4 A8 A5 60 AF B2 A9 A7 A9 6D  0  0",
    " 0 AE A1 AC 60 B6 A1 AC B5 A5 7B 60 B4 A8 A5 AE 60 B4 A8 A5 60 A3 A1 AC A3 B5 AC A1 B4 A5 A4  0",
    " 0 B6 A1 AC B5 A5 60 A5 B8 A1 A3 B4 AC B9 60 71 60 A3 B9 A3 AC A5 60 AC A1 B4 A5 B2 6E  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0 36 45 52 49 46 59 49 4E 47  0 4F 50 45 4E  0 42 55 53  0 42 45 48 41 56 49 4F 52  E  0  0  0",
    " 0 60 60 60 60 60 60 97 6D 60 97 6D 60 97 92 60 97 6D 60 97 6D 60 97 6D 60 97 6D 60 97 92  0  0",
    " 0 72 70 70 70 6B 60 70 60 60 71 60 60 72 60 60 73 60 60 74 60 60 75 60 60 76 60 60 77 60  0  0",
    " 0 60 60 92 70 7A C0 D0 CD C0 D0 CD C0 D0 D0 C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 D0  0  0",
    " 0 60 60 92 71 7A C0 D0 CD C0 D0 CD C0 D0 D0 C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 D0  0  0",
    " 0 60 60 92 73 7A C0 D0 CD C0 D0 CD C0 D0 D0 C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 D0  0  0",
    " 0 60 60 92 75 7A C0 D0 CD C0 D0 CD C0 D0 D0 C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 D0  0  0",
    " 0 60 60 92 76 7A C0 D0 CD C0 D0 CD C0 D0 D0 C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 CD C0 D0 D0  0  0",
    " 0 2F 2B 1B  0 36 45 52 49 46 59 49 4E 47  0 4F 50 43 4F 44 45 53  E  E  E  0  0  0  0  0  0  0",
    " 0 D0 E5 D2 E5 D4 E5 D6 E5 E3 E5 E5 E5 C0 D1 E5 D3 E5 D5 E5 D7 E5 E4 E5 E6 E5 C0  0  0  0  0  0",
    " 0 D0 E6 D2 E6 D4 E6 D6 E6 E3 E6 E5 E6 C0 D1 E6 D3 E6 D5 E6 D7 E6 E4 E6 E6 E6 C0  0  0  0  0  0",
    " 0 D0 D3 D2 D3 D4 D3 D6 D3 E3 D3 E5 D3 C0 D1 D3 D3 D3 D5 D3 D7 D3 E4 D3 E6 D3 C0  0  0  0  0  0",
    " 0 D1 E2 D3 E2 D5 E2 D7 E2 E4 E2 E6 E2 C0 C0 C0 C0 C0 C0 C0 C0 C0 C0 C0 C0 C0 C0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0 30 41 53 53 45 44  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
]
